tests/e2e/e2e_test.go

```diff
@@ -111,3 +109,257 @@ func (s *IntegrationTestSuite) TestBankTokenTransfer() {
+		)
+	})
+}
+
+func (s *IntegrationTestSuite) submitLegacyProposalFundGovAccount(chainAAPIEndpoint string, sender string) {
+	s.Run("submit_legacy_community_spend_proposal_to_fund_gov_acct", func() {
+		s.execGovSubmitLegacyGovProposal(s.chainA, 0, chainAAPIEndpoint, sender, "/root/.gaia/config/proposal.json", fees.String(), "community-pool-spend")
+
+		s.Require().Eventually(
+			func() bool {
+				proposal, err := queryGovProposal(chainAAPIEndpoint, 1)
+				s.Require().NoError(err)
+
+				return (proposal.GetProposal().Status == govv1beta1.StatusDepositPeriod)
+			},
+			15*time.Second,
+			5*time.Second,
+		)
+	})
+}
+
+func (s *IntegrationTestSuite) submitNewGovProposal(chainAAPIEndpoint string, sender string, proposalId int, proposalPath string) {
+	s.Run("submit_new_gov_proposal", func() {
+		s.execGovSubmitProposal(s.chainA, 0, chainAAPIEndpoint, sender, proposalPath, fees.String())
+
+		s.Require().Eventually(
+			func() bool {
+				proposal, err := queryGovProposal(chainAAPIEndpoint, proposalId)
+				s.T().Logf("Proposal: %s", proposal.String())
+				s.Require().NoError(err)
+
+				return (proposal.GetProposal().Status == govv1beta1.StatusDepositPeriod)
+			},
+			15*time.Second,
+			5*time.Second,
+		)
+	})
+}
+
+func (s *IntegrationTestSuite) depositGovProposal(chainAAPIEndpoint string, sender string, proposalId int) {
+	s.Run("deposit_gov_proposal", func() {
+		s.execGovDepositProposal(s.chainA, 0, chainAAPIEndpoint, sender, proposalId, depositAmount, fees.String())
+
+		s.Require().Eventually(
+			func() bool {
+				proposal, err := queryGovProposal(chainAAPIEndpoint, proposalId)
+				s.Require().NoError(err)
+
+				return (proposal.GetProposal().Status == govv1beta1.StatusVotingPeriod)
+			},
+			15*time.Second,
+			5*time.Second,
+		)
+	})
+}
+
+func (s *IntegrationTestSuite) voteGovProposal(chainAAPIEndpoint string, sender string, proposalId int, vote string, weighted bool) {
+	s.Run("vote_gov_proposal", func() {
+		if weighted {
+			s.execGovWeightedVoteProposal(s.chainA, 0, chainAAPIEndpoint, sender, proposalId, vote, fees.String())
+		} else {
+			s.execGovVoteProposal(s.chainA, 0, chainAAPIEndpoint, sender, proposalId, vote, fees.String())
+		}
+
+		s.Require().Eventually(
+			func() bool {
+				proposal, err := queryGovProposal(chainAAPIEndpoint, proposalId)
+				s.Require().NoError(err)
+
+				return (proposal.GetProposal().Status == govv1beta1.StatusPassed)
+			},
+			15*time.Second,
+			5*time.Second,
+		)
+	})
+}
+
+func (s *IntegrationTestSuite) verifyChainHaltedAtUpgradeHeight(c *chain, valIdx int, upgradeHeight int) {
+	s.Require().Eventually(
+		func() bool {
+			currentHeight := s.getLatestBlockHeight(c, valIdx)
+
+			return currentHeight == upgradeHeight
+		},
+		30*time.Second,
+		5*time.Second,
+	)
+
+	counter := 0
+	s.Require().Eventually(
+		func() bool {
+			currentHeight := s.getLatestBlockHeight(c, valIdx)
+
+			if currentHeight > upgradeHeight {
+				panic("Chain did not halt at upgrade height")
```

Comment:
should this panic or return false until the timer runs out forcing the test framework to register the error? the message could be returned as the fail mode of the actual test right?